pisa/population.py

```diff
@@ -0,0 +1,206 @@
+
+    def get_population_data_with_selected_method(self) -> pd.DataFrame:
+        """Select correct method according to the data_source given and return the population data as a DataFrame."""
+        if self.data_source == self.DataSource.WORLD_POP.value:
+            return self.get_population_worldpop()
+        elif self.data_source == self.DataSource.FACEBOOK.value:
+            return self.get_population_facebook()
+        else:
+            raise ValueError(
+                f"Invalid data source: {self.data_source}. Must be one of {[ds.value for ds in Population.DataSource]}.")
+
+    @staticmethod
+    def group_population(population_df: pd.DataFrame, population_resolution: int) -> GeoDataFrame:
+        """ Group population data by longitude and latitude and return as a GeoDataFrame."""
+        population_df.loc[:, ["longitude", "latitude"]] = population_df[["longitude", "latitude"]].round(
+            population_resolution)
+
+        population = (
+            population_df.groupby(["longitude", "latitude"], as_index=False)["population"]
+            .sum()
+            .reset_index(names="ID")
+        )
+        population["population"] = population["population"].round(2)
+        population = GeoDataFrame(population, geometry=points_from_xy(population.longitude, population.latitude))
+        return population
+
+    def get_population_facebook(self) -> pd.DataFrame:
+        """Download & process data from the chosen datasource 'facebook'. Returns a DataFrame with population data."""
+
+        downloaded_data = self.download_population_facebook(
+            iso3_country_code=self.iso3_country_code,
+        )
+
+        processed_data = self.process_population_facebook(
+            downloaded_data,
+            iso3_country_code=self.iso3_country_code,
+            admin_area_boundaries=self.admin_area_boundaries,
+        )
+
+        return processed_data
+
+    @staticmethod
+    def download_population_facebook(
+        iso3_country_code: str
+    ) -> pd.DataFrame:
+        """Download population data from 2020 from facebook for a country defined by the iso3_country_code."""
+        try:
+            Configuration.create(
+                hdx_site="prod", user_agent="Get_Population_Data", hdx_read_only=True
+            )
+        except Exception as e:
+            print(f"Warning: Unable to configure HDX. Error: {e}")
+
+        resource = Resource.search_in_hdx(
+            f"name:{iso3_country_code.lower()}_general_2020_csv.zip"
+        )
+        if not resource:
+            raise ValueError(f"No resource found for country code: {iso3_country_code}")
+
+        url = resource[0]["download_url"]
+        filehandle, _ = urllib.request.urlretrieve(url)
+
+        df = pd.read_csv(filehandle, compression="zip")
+        return df
+
+    @staticmethod
+    def process_population_facebook(downloaded_data: pd.DataFrame, iso3_country_code: str, admin_area_boundaries: Polygon | MultiPolygon) -> pd.DataFrame:
```

Comment:
I see that both the facebook and worldpop class need the iso3_country_code. facebook also needs the admin area boundaries.

instead of having the user passing these two arguments, would it make sense to pass the an admin area object to when initializing the `Population` class? then the worldpop and facebook methods could just call the methods that are already built in `Population` to get the country code and the admin area boundaries.

What do you think?